homeassistant/bootstrap.py

```diff
@@ -102,56 +102,68 @@ def _async_setup_component(hass: core.HomeAssistant,
     """
     # pylint: disable=too-many-return-statements,too-many-branches
     # pylint: disable=too-many-statements
+    if not hasattr(hass, 'setup_lock'):
+        hass.setup_lock = asyncio.Lock(loop=hass.loop)
+
     if domain in hass.config.components:
         return True
 
-    if domain in _CURRENT_SETUP:
-        _LOGGER.error('Attempt made to setup %s during setup of %s',
-                      domain, domain)
-        return False
+    did_lock = False
+    if not hass.setup_lock.locked():
+        yield from hass.setup_lock.acquire()
+        did_lock = True
 
-    config = yield from async_prepare_setup_component(hass, config, domain)
+    try:
```

Comment:
yield from context managers is Python 3.5 only.

RLock won't work because everything runs in 1 thread with event loop. That's why the check logic had to move to the places were we schedule setup_component -> discovery.